src/packets.toit

```diff
@@ -135,100 +228,140 @@ class ConnectPacket extends Packet:
+  static IDENTIFIER_REJECTED ::= 0x02
+  static SERVER_UNAVAILABLE ::= 0x03
+  static BAD_USERNAME_OR_PASSWORD ::= 0x04
+  static NOT_AUTHORIZED ::= 0x05
 
-  constructor:
-    return_code = 0
+  return_code /int
+  session_present /bool
+
+  constructor --.return_code=0 --.session_present=false:
     super TYPE
 
-  constructor.deserialize reader/reader.BufferedReader:
+  constructor.deserialize_ reader/reader.BufferedReader:
     data := reader.read_bytes 2
+    session_present = data[0] & 0x01 != 0
     return_code = data[1]
     super TYPE
 
   variable_header -> ByteArray:
-    return #[0, 0]
+    return #[ session_present ? 1 : 0, return_code]
 
   payload -> ByteArray: return #[]
 
 class PublishPacket extends Packet:
   static TYPE ::= 3
 
-  topic/string
-  payload/ByteArray
-  qos/int
-  packet_id/int?
+  topic /string
+  payload /ByteArray
+  packet_id /int?
 
-  constructor.deserialize reader/reader.BufferedReader size/int flags/int:
-    qos = (flags >> 1) & 0b11
-    data := reader.read_bytes 2
-    topic_length := binary.BIG_ENDIAN.uint16 data 0
-    topic = reader.read_string topic_length
-    size -= 2 + topic_length
+  constructor.deserialize_ reader/reader.BufferedReader size/int flags/int:
+    retain := flags & 0b1 != 0
+    qos := (flags >> 1) & 0b11
+    duplicate := flags & 0b1000 != 0
+    topic = Packet.decode_string reader
+    size -= 2 + topic.size
     if qos > 0:
-      data = reader.read_bytes 2
-      packet_id = binary.BIG_ENDIAN.uint16 data 0
+      packet_id = Packet.decode_uint16 reader
       size -= 2
     else:
       packet_id = null
     payload = reader.read_bytes size
-    super TYPE
+    super TYPE --flags=(qos << 1) | (retain ? 1 : 0) | (duplicate ? 0b1000 : 0)
 
-  constructor .topic .payload --.qos/int --retain --.packet_id:
+  constructor .topic .payload --qos/int --retain/bool --.packet_id --duplicate=false:
     super TYPE
-      --flags=(qos << 1) | (retain ? 1 : 0)
+        --flags=(qos << 1) | (retain ? 1 : 0) | (duplicate ? 0b1000 : 0)
 
   variable_header -> ByteArray:
     buffer := bytes.Buffer
     Packet.encode_string buffer topic
-    if packet_id:
-      data := ByteArray 2
-      binary.BIG_ENDIAN.put_uint16 data 0 packet_id
-      buffer.write data
+    if packet_id: buffer.write (Packet.encode_uint16 packet_id)
     return buffer.bytes
 
-class PubAckPacket extends Packet implements PacketIDAck:
+  retain -> bool: return flags & 0b1 != 0
+  qos -> int: return (flags >> 1) & 0b11
+  duplicate -> bool: return flags & 0b1000 != 0
+
+  /**
+  Creates a new packet with the provided parameters replacing the current values.
+
+  Optional arguments don't work if one wants to pass in null. As such, the
+    $packet_id must be given as integer. If the packet-id should be null, then '-1' should
+    be used instead.
+  */
+  with -> PublishPacket
+      --topic/string?=null
+      --payload/ByteArray?=null
+      --qos/int?=null
+      --retain/bool?=null
+      --packet_id/int=-2
```

Comment:
Use named constant for -2?

Reply:
I will change the code so there are two `with` functions. One with, and one without the `--packet_id`. This way I can avoid this hack.